JavaScript: Add support for state-based taint tracking.

xiemaisi opened the pull request:
This is still quite rough, so I'm marking it as WIP, but perhaps opening a PR is the best way to get the discussion started.

The approach I've taken is to introduce an abstract subclass of `string` called `FlowLabel` whose elements represent taint kinds/states. There are two standard labels "data" and "taint", the former describing values that directly originate from a flow source, the latter values that are derived from a flow source via one or more transformations (such as string operations).

Individual queries can define additional flow labels. Flow configurations can associate sources and sinks with labels, in the former case indicating a label to attach to any value arising from the source, and in the latter case requiring that values have a certain label in order to be consumed by the sink. Similarly, sanitisers can be associated with labels to selectively block flow of values with a certain label but not others. Finally, additional flow steps can specify an input and output label, which correspond to a state transition.

By default, sources produce values labelled with "data" and sinks consume values with any label, value-preserving flow steps map "data" to "data", and taint steps map both "data" and "taint" to "taint". These defaults allow us to reuse existing configurations without any changes.

As an example of the use of this machinery, I've updated the client-side URL redirect query to use a custom flow label representing `document.url`. Reading the `query` or `hash` property of a value labelled in this way and removing the leading `?` or `#` character changes the label to `taint`.

I've evaluated these changes on `default.slugs`, showing a very modest [slowdown](https://git.semmle.com/gist/max/dcaa3574a918ce1b05dcf50eda9b4652) (internal link) as one would expect, and no changes in results.

asger-semmle commented on javascript/ql/src/semmle/javascript/dataflow/Configuration.qll:
```diff
@@ -350,15 +412,21 @@ private predicate exploratoryFlowStep(DataFlow::Node pred, DataFlow::Node succ,
-  cfg.isSink(nd) or nd.(AdditionalSink).isSinkFor(cfg)
+private predicate isSink(DataFlow::Node nd, DataFlow::Configuration cfg, FlowLabel lbl) {
+  (cfg.isSink(nd) or nd.(AdditionalSink).isSinkFor(cfg)) and
+  lbl = any(FlowLabel f)
```
As discussed offline, this seems a little risky, at least superficially.

It will probably be inlined anyway, but if it one day stops being inlined, this could be quite costly for configurations that just specify `any()` as sink.

xiemaisi replied:
Updated to restrict it to `any(StandardFlowLabel lbl`).

esben-semmle commented on javascript/ql/src/semmle/javascript/security/dataflow/ClientSideUrlRedirect.qll:
```diff
@@ -35,24 +43,37 @@ module ClientSideUrlRedirect {
 
+    override predicate isSource(DataFlow::Node source, DataFlow::FlowLabel lbl) {
+      source instanceof RemoteFlowSource and
+      lbl = DataFlow::FlowLabel::taint()
```
Why does `RemoteFlowSource` start out as being of kind "taint"? I would have expected "data".

I see that the rest of the configuration expects "taint", but it should be able to handle both "data" and "taint" IMO.

xiemaisi replied:
Yes, "data" would be better. I think the original reason for making it "taint" was that `isSink` expects "taint", but we should instead make that accept "data" as well.

esben-semmle commented on javascript/ql/src/semmle/javascript/security/dataflow/ClientSideUrlRedirect.qll:
```diff
@@ -84,39 +105,13 @@ module ClientSideUrlRedirect {
-        queryAccess(nd, this)
-      )
-    }
+  abstract class UrlSink extends DataFlow::Node {
```
Missing docstring.

esben-semmle commented on javascript/ql/test/library-tests/InterProceduralFlow/TaintTracking.expected:
```diff
@@ -36,6 +36,9 @@
 | tst2.js:3:17:3:26 | "tainted2" | tst2.js:11:15:11:24 | g(source2) |
 | tst2.js:6:24:6:37 | "also tainted" | tst2.js:10:15:10:24 | g(source1) |
 | tst2.js:6:24:6:37 | "also tainted" | tst2.js:11:15:11:24 | g(source2) |
+| tst4.js:2:16:2:24 | "tainted" | tst4.js:15:15:15:31 | id(still_tainted) |
```
just checking: these tests behave identically on the current `ql/master`, right?

xiemaisi replied:
I think so, yes. This is a regression test for a mistake I made while working on this PR but later fixed.

esben-semmle commented on javascript/ql/src/semmle/javascript/dataflow/Configuration.qll:
```diff
@@ -220,11 +220,14 @@ abstract class FlowLabel extends string {
+ * A standard flow label, that is, either `FlowLabel::data()` or `FlowLabel::taint()`.
+ */
+private class StandardFlowLabel extends FlowLabel {
+  StandardFlowLabel() { this = "data" or this = "taint" }
```
Minor design issue on internal class: shouldn't this be using `FlowLabel::data()`  and `FlowLabel::taint()` to hide the string implementation detail?

xiemaisi replied:
That would lead to an empty recursion due to type-joining on the result type of `FlowLabel::data()` and `FlowLabel::taint()`.

xiemaisi commented:
> This deserves a prominent change-note, and perhaps a ping to the other language teams.

Absolutely. I'll schedule a meeting with the other language maintainers some time next week.

xiemaisi commented on javascript/ql/src/semmle/javascript/dataflow/internal/FlowSteps.qll:
```diff
@@ -301,13 +309,24 @@ module PathSummary {
+  private PathSummary mkPathSummary(boolean hasCall, boolean hasReturn, Boolean valuePreserving) {
+    exists (FlowLabel start, FlowLabel end |
+      if valuePreserving = false then
+        end = FlowLabel::taint()
```
This was changed later.

xiemaisi commented:
I've added `TaintKind` as an alias for `FlowLabel`, so we can use whichever is clearer in context. I've also added a brief change note, to be supplemented by a fuller description in the library overview (in a separate PR, since that's still in the internal repo).
